File: MATLAB/Discretization/obst_zero_val.m
%==========================================================================
function [val] = obst_zero_source(pos, val, obst)
%--------------------------------------------------------------------------
% Set value to zero inside obstacle.
% 
% pos  - position of the variable, C, X, Y or Z
% val  - value to be zeroed in obstacle
% obst - matrix holding positions of obstacle
%--------------------------------------------------------------------------

if pos==C  
  val = val .* not(obst); 
  
elseif pos==X  
  obst_x = max(obst(1:end-1,:,:), obst(2:end,:,:));
  val = val .* not(obst_x); 
  
elseif pos==Y  
  obst_y = max(obst(:,1:end-1,:), obst(:,2:end,:));
  val = val .* not(obst_y); 
  
elseif pos==Z  
  obst_z = max(obst(:,:,1:end-1), obst(:,:,2:end));
  val = val .* not(obst_z); 
end

end

File: PYTHON/scrins/discretization/obst_zero_val.py
from scipy import logical_not as lnot
from scipy import maximum as mx

from scrins.constants.coordinates import X, Y, Z
from scrins.constants.compass import C

#==========================================================================
def obst_zero_val(d, val, obst):
#--------------------------------------------------------------------------
# Set value to zero inside obstacle.
# 
# d    - position of the variable, C, X, Y or Z
# val  - value to be zeroed in obstacle
# obst - matrix holding positions of obstacle
#--------------------------------------------------------------------------

  if d == C:  
    val = val * lnot(obst)
    
  elif d==X:
    obst_x = mx(obst[:-1,:,:], obst[1:,:,:])
    val = val * lnot(obst_x)
    
  elif d==Y: 
    obst_y = mx(obst[:,:-1,:], obst[:,1:,:])
    val = val * lnot(obst_y)
    
  elif d==Z: 
    obst_z = mx(obst[:,:,:-1], obst[:,:,1:])
    val = val * lnot(obst_z)

  return val  # end of function